db/src/models/schedule.rs

```diff
@@ -0,0 +1,155 @@
+pub struct Schedule {
+    pub id: Uuid,
+    pub game_id: Uuid,
+    pub proposer_id: Uuid,
+    pub start_t: DateTime<Utc>,
+    pub agreed: bool,
+}
+
+impl Schedule {
+    pub async fn accept(
+        &mut self,
+        user_id: Uuid,
+        conn: &mut DbConn<'_>,
+    ) -> Result<GameId, DbError> {
+        let game: Game = games::table.find(self.game_id).get_result(conn).await?;
+        if !self.is_proposer(user_id) && is_player(user_id, &game) {
+            //unset all schedules for this game
+            diesel::update(game_schedules::table.filter(game_schedules::game_id.eq(self.game_id)))
+                .set(game_schedules::agreed.eq(false))
+                .execute(conn)
+                .await?;
+            //set this schedule only
+            diesel::update(game_schedules::table.find(self.id))
+                .set(game_schedules::agreed.eq(true))
+                .execute(conn)
+                .await?;
+            self.agreed = true;
+            Ok(GameId(game.nanoid))
+        } else {
+            Err(DbError::Unauthorized)
+        }
+    }
+
+    pub async fn insert(&self, conn: &mut DbConn<'_>) -> Result<usize, DbError> {
+        Ok(diesel::insert_into(game_schedules::table)
+            .values((
+                game_schedules::id.eq(self.id),
+                game_schedules::game_id.eq(self.game_id),
+                game_schedules::proposer_id.eq(self.proposer_id),
+                game_schedules::start_t.eq(self.start_t),
+                game_schedules::agreed.eq(false),
+            ))
+            .execute(conn)
+            .await?)
+    }
+
+    pub async fn cancel(
+        &mut self,
+        user_id: Uuid,
+        conn: &mut DbConn<'_>,
+    ) -> Result<GameId, DbError> {
+        let game: Game = games::table.find(self.game_id).get_result(conn).await?;
+        if self.is_proposer(user_id) && is_player(user_id, &game) {
+            diesel::delete(game_schedules::table.find(self.id))
+                .execute(conn)
+                .await?;
+            Ok(GameId(game.nanoid))
+        } else {
+            Err(DbError::Unauthorized)
+        }
+    }
+
+    pub async fn reject(
+        &mut self,
+        user_id: Uuid,
+        conn: &mut DbConn<'_>,
+    ) -> Result<GameId, DbError> {
+        let game: Game = games::table.find(self.game_id).get_result(conn).await?;
+        if !self.is_proposer(user_id) && is_player(user_id, &game) {
+            diesel::delete(game_schedules::table.find(self.id))
+                .execute(conn)
+                .await?;
+            Ok(GameId(game.nanoid))
+        } else {
+            Err(DbError::Unauthorized)
+        }
+    }
+
+    pub async fn get_first_agreed(game_id: Uuid, conn: &mut DbConn<'_>) -> Result<Self, DbError> {
+        Ok(game_schedules::table
+            .filter(game_schedules::game_id.eq(game_id))
+            .filter(game_schedules::agreed.eq(true))
+            .filter(game_schedules::start_t.gt(now))
+            .first(conn)
```

Comment:
Is this deterministic? Will we get the same results if we run this function twice? I actually don't know...

Reply:
Likely not totally deterministic, im trying to avoid sending expired schedules but there is also filtering on the client side so this can be removed if you want